File: simulation.go
package main

import (
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"runtime"
	"runtime/pprof"
	"time"
)

var (
	threadNum int = runtime.NumCPU()
	height        = 6
	reps          = threadNum
)

func main() {
	runtime.GOMAXPROCS(threadNum)
	var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
	flag.Parse()
	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal(err)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	localRand := rand.New(rand.NewSource(time.Now().UnixNano()))
	g := NewGrid()
	g.PlaceRandom(localRand)
	g.PlaceRandom(localRand)

	fmt.Printf("%s\n", g)

	moves := 0

	for {
		avgBest := g.NextMove(height, reps, threadNum)
		if avgBest == -1 {
			fmt.Println("Couldn't find good direction. Game over")
			break
		}

		fmt.Printf("Moving %s\n", DirectionToString(avgBest))
		g.Move(avgBest)
		moves++

		if !g.PlaceRandom(localRand) {
			fmt.Println("Couldn't place piece. Game over")
			break
		}

		fmt.Printf("%s\n", g)

	}

	fmt.Printf("After %d moves:\n", moves)
	fmt.Println(g)
}

File: simulation.cpp
#include "grid.hpp"
#include "ai.hpp"

#include <iostream>

const size_t height = 6;
const size_t thread_num = std::thread::hardware_concurrency();
const size_t reps = thread_num*2;

int main() {
    Grid g;
    size_t moves = 0;
    g.place_random();
    g.place_random();

    std::cout << g << std::endl;

    while (true) {
        int best_move = next_move(g, height, reps, thread_num);
        if (best_move == -1) {
            std::cout << "Couldn't find good direction. Game over" << std::endl;
			break;
        }

        std::cout << "Moving " << direction_to_string(best_move) << std::endl;
        g.move(best_move);
        moves++;

        if (!g.place_random()) {
            std::cout << "Couldn't place piece. Game over" << std::endl;
			break;
        }

        std::cout << g << std::endl;
    }

    std::cout << "After " << moves << " moves:" << std::endl;
    std::cout << g << std::endl;
}
